src/Configuration/Config.php

```diff
@@ -18,40 +21,109 @@ class Config
+        $this->cache = new FilesystemCache();
         $this->pathResolver = new PathResolver(dirname(dirname(__DIR__)), []);
+        $this->data = $this->getConfig();
+    }
+
+    private function getConfig(): Collection
+    {
+        $this->stopwatch->start('bolt.parseconfig');
+
+        if ($this->validCache()) {
+            $data = $this->getCache();
+        } else {
+            $data = $this->parseConfig();
+            $this->setCache($data);
+        }
+
+        $this->stopwatch->stop('bolt.parseconfig');
 
-        $this->data = $this->parseConfig();
+        return $data;
+    }
+
+    private function validCache(): bool
+    {
+        if (! $this->cache->has('config_cache') || ! $this->cache->has('config_timestamps')) {
+            return false;
+        }
+
+        $timestamps = $this->cache->get('config_timestamps');
+
+        foreach ($timestamps as $filename => $timestamp) {
+            if (filemtime($filename) > $timestamp) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private function getCache(): Collection
+    {
+        return $this->cache->get('config_cache');
+    }
+
+    private function setCache($data): void
+    {
+        $this->cache->set('config_cache', $data);
     }
 
     /**
      * Load the configuration from the various YML files.
      */
-    public function parseConfig(): Collection
+    private function parseConfig(): Collection
     {
         $general = new GeneralParser();
 
         $config = collect([
             'general' => $general->parse(),
         ]);
 
-        $this->data = $config;
-
         $taxonomy = new TaxonomyParser();
         $config['taxonomies'] = $taxonomy->parse();
 
-        $contentTypes = new ContentTypesParser($this->get('general')['accept_file_types']);
+        $contentTypes = new ContentTypesParser($config->get('general')['accept_file_types']);
         $config['contenttypes'] = $contentTypes->parse();
 
+        // @todo Add these config files if needed, or refactor them out otherwise
         //'menu' => $this->parseConfigYaml('menu.yml'),
         //'routing' => $this->parseConfigYaml('routing.yml'),
         //'permissions' => $this->parseConfigYaml('permissions.yml'),
         //'extensions' => $this->parseConfigYaml('extensions.yml'),
 
+        $this->getConfigFilesTimestamps($general, $taxonomy, $contentTypes);
+
         return $config;
     }
 
+    private function getConfigFilesTimestamps(BaseParser ...$configs): void
+    {
+        $timestamps = [];
+
+        foreach ($configs as $config) {
+            foreach ($config->getFilenames() as $file) {
+                $timestamps[$file] = filemtime($file);
+            }
+        }
+
+        $envFilename = dirname(dirname(__DIR__)) . '/.env';
+        if (file_exists($envFilename)) {
+            $timestamps[$envFilename] = filemtime($envFilename);
+        }
+
+        $this->cache->set('config_timestamps', $timestamps);
```

Comment:
$this->cache is not yet set here, that's why parseConfigFiles should return it too

Reply:
Done. :-)